server-adapters/resteasy-netty4/src/main/java/org/jboss/resteasy/plugins/server/netty/MultiPromise.java

```diff
@@ -0,0 +1,55 @@
+package org.jboss.resteasy.plugins.server.netty;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+public class MultiPromise
+{
+
+   private ChannelHandlerContext ctx;
+   private AtomicInteger created = new AtomicInteger();
+   private AtomicReference<Throwable> cause = new AtomicReference<>();
+   private ChannelPromise promise;
+   private boolean started;
+
+   public MultiPromise(final ChannelHandlerContext ctx, final ChannelPromise promise)
+   {
+      this.ctx = ctx;
+      this.promise = promise;
+   }
+
+   public ChannelPromise newPromise()
+   {
+      created.getAndIncrement();
+      return ctx.newPromise().addListener(f -> {
+         if(!f.isSuccess())
+            cause.compareAndSet(null, f.cause());
+         if(created.decrementAndGet() == 0)
```

Comment:
If this happens as 'readyToForward' is being called both this method and 'readyToForward' could see the count as zero and the logic in forward() would be called twice.

Reply:
It can be called twice, because the output promise is checked for `isDone` to not complete it twice.